### PR TITLE
Collect errors and set non-zero exit code if any occured

### LINKED ISSUE #201: Provide an option to only validate x-checker-data
- [x] Some option like `--validate-only` that would cause f-e-d-c to only check if `x-checker-data` is valid, without doing actual updates and exiting with non-zero status only when an error occurred, could be useful to integrate f-e-d-c as part of Flathub CI.
- [ ] Preferably, f-e-d-c in `--validate-only` mode shouldn't hit the network, but it would require a bit more work (we'll need to split some `load_checker_data()` method from `check()` in each checker). But this can be done afterwards.

### PR DESCRIPTION
**Pull request by gasinvein**
Collect all the non-fatal errors occurred during check, and if a special option is given, set the exit status based on amount of errors collected.
Fixes #201

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -285,8 +292,8 @@ def parse_cli_args(cli_args=None):
         action="store_true",
     )
     parser.add_argument(
-        "--validate-only",
-        help="Do not apply updates, only check if metadata is valid",
+        "--validate",
+        help="Set exit code based on errors amount",
```
Comment (wjt) [inline src/main.py]: Why is this conditional? Why not always exit with a defined error code in this case? (Sincere question.)

```
         help="Exit with non-zero status if errors are found while checking",
```

Comment (gasinvein) [reply src/main.py]: Currently we set exit code based on amount of outdated data - 0 if all up to date and 1 if something is outdated. I suppose we should retain this default behaviour?

Comment (gasinvein) [reply src/main.py]: Maybe we should combine non-zero exit code? E.g. `1` if outdated, `2` if errors, `3` if both.

Comment (gasinvein) [reply src/main.py]: Now this is unconditional.

**Comment by gasinvein** [comment]
@wjt Will this work for Endless' CI, given the default behaviour has changed? I guess we may need to add the new `--check-outdated` option to the wrapper?

File: src/main.py [hunk 2]
```diff
@@ -298,41 +306,50 @@ def parse_cli_args(cli_args=None):
...
-                        open_pr(
-                            subject, body, branch, manifest_checker=manifest_checker
-                        )
-                return (outdated_num, True)
+    return outdated_num, errors_num, did_update
 
-            log.warning("Can't automatically fix any of the above issues")
 
-    return (outdated_num, False)
+class ResultCode(IntFlag):
```
Comment (wjt) [inline src/main.py:341]: I have never seen IntFlag before. Nice!

File: src/main.py [hunk 3]
```diff
@@ -298,41 +306,50 @@ def parse_cli_args(cli_args=None):
...
+    SUCCESS = 0
+    ERROR = 1
+    OUTDATED = 2
 
 
 def main():
+    res = ResultCode(0)
```
Comment (wjt) [inline src/main.py]: Why not:

```suggestion
    res = ResultCode.SUCCESS
```

?

File: src/main.py [hunk 4]
```diff
@@ -104,6 +106,12 @@ def print_outdated_external_data(manifest_checker: checker.ManifestChecker):
     return len(ext_data)
 
 
+def print_errors(manifest_checker: checker.ManifestChecker):
```
Comment (wjt) [inline src/main.py]:
```suggestion
def print_errors(manifest_checker: checker.ManifestChecker) -> int:
```

A little weird that a function called `print_*` returns a value and doesn't print :-)

**Review by wjt — APPROVED** [review]
Looks good. I don't think the Endless CI needs to change.

File: src/main.py (hunk 4 above)
Comment (gasinvein) [reply src/main.py]: It returns an `int` just like `print_outdated_external_data()` does.
I didn't yet figure out how to print errors in a readable and useful manner.

File: src/main.py (hunk 2 above)
Comment (gasinvein) [reply src/main.py:341]: Very bitwise, much C

**Pull request merged**